### PR TITLE
feat: adding the annotaTR tool

### LINKED ISSUE #209: Converting Beagle imputed VCF for TRTools
Hi

First of all, thanks for developing TRTools! I have a question regarding imputed STR input files.

I have a Beagle (v5.4) imputed VCF file that looks like below:

>##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##filedate=20230503
##source="beagle.22Jul22.46e.jar"
##INFO=<ID=AF,Number=A,Type=Float,Description="Estimated ALT Allele Frequencies">
##INFO=<ID=DR2,Number=A,Type=Float,Description="Dosage R-Squared: estimated squared correlation between estimated REF dose [P(RA) + 2*P(RR)] and true REF dose">
##INFO=<ID=IMP,Number=0,Type=Flag,Description="Imputed marker">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=DS,Number=A,Type=Float,Description="estimated ALT dose [P(RA) + 2*P(AA)]">
##contig=<ID=1>
##contig=<ID=2>
##contig=<ID=3>
##contig=<ID=4>
##contig=<ID=5>
##contig=<ID=6>
##contig=<ID=7>
##contig=<ID=8>
##contig=<ID=9>
##contig=<ID=10>
##contig=<ID=11>
##contig=<ID=12>
##contig=<ID=13>
##contig=<ID=14>
##contig=<ID=15>
##contig=<ID=16>
##contig=<ID=17>
##contig=<ID=18>
##contig=<ID=19>
##contig=<ID=20>
##contig=<ID=21>
##contig=<ID=22>

| #CHROM | POS    | ID         | REF                    | ALT                                                                                                                                                                                                                                     | QUAL | FILTER | INFO                                                                                          | FORMAT | Sample1               |
|--------|--------|------------|------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------|--------|-----------------------------------------------------------------------------------------------|--------|-----------------------|
| 18     | 136703 | STR_614109 | TCCGGCAAAAAAAAAAAAAAAA | TCCAGCAAAAAAAAAAAAAAAA,TCCGGAAAAAAAAAAAAAAAAA,TGCGGCAAAAAAAAAAAAAAAA,TCCGGCAAAAAAAAAAAAAAAAA,TCCGGAAAAAAAAAAAAAAAAAAGAA,TCCGGCAAAAAAAAAAAAAAAAAGAA,TCCGGAAAAAAAAAAAAAAAAAAAGAA,TCCGGCAAAAAAAAAAAAAAAAAAGAA,TCCGGCAAAAAAAAAAAAAAAAAAAGAA | .    | PASS   | DR2=0,0.35,0,0.28,0,0.97,0.33,0.93,0.45;AF=0,0.008,0,0.0082,0,0.0352,0.0044,0.2689,0.0045;IMP | GT:DS  | 0\|8:0,0,0,0,0,0,0,1,0 |

The imputation reference panel is from: http://gymreklab.com/2018/03/05/snpstr_imputation.html that was described in the Saini et al. paper: https://www.nature.com/articles/s41467-018-06694-0

I typically split these STR alleles and perform biallelic STR GWAS on binary phenotypes of interest, however I am also interested in a length-based STR GWAS. That is how I came across with associaTR available in TRTools, that seems to be capable of running length-based GWAS. I made a test as below with chr18 STRs:

`associaTR results.tsv input_chr18_fortesting.vcf.gz cases phenotypes_forTest_caseControl.npy --same-samples --beagle-dosages --vcftype hipstr`

...however this gave an error as:

>Traceback (most recent call last):
  File "/home/fahri/miniconda3/bin/associaTR", line 10, in <module>
    sys.exit(run())
  File "/home/fahri/miniconda3/lib/python3.9/site-packages/trtools/associaTR/associaTR.py", line 582, in run
    main(args)
  File "/home/fahri/miniconda3/lib/python3.9/site-packages/trtools/associaTR/associaTR.py", line 599, in main
    perform_gwas(
  File "/home/fahri/miniconda3/lib/python3.9/site-packages/trtools/associaTR/associaTR.py", line 449, in perform_gwas
    perform_gwas_helper(
  File "/home/fahri/miniconda3/lib/python3.9/site-packages/trtools/associaTR/associaTR.py", line 207, in perform_gwas_helper
    extra_detail_fields = next(genotype_iter)
  File "/home/fahri/miniconda3/lib/python3.9/site-packages/trtools/associaTR/load_and_filter_genotypes.py", line 117, in load_trs
    inferred_vcftype = trh.InferVCFType(vcf, vcftype if vcftype else 'auto')
  File "/home/fahri/miniconda3/lib/python3.9/site-packages/trtools/utils/tr_harmonizer.py", line 209, in InferVCFType
    raise TypeError('Could not identify the type of this vcf')
TypeError: Could not identify the type of this vcf. 

Of note, above I set --vcftype as hipstr, as Saini et al. paper seems to use HipSTR.

Then I realized this section of the documentation regarding Beagle: https://trtools.readthedocs.io/en/stable/CALLERS.html#beagle and used trtools_prep_beagle_vcf.sh convert my imputed VCF into a VCF that can be used in associaTR:

`trtools_prep_beagle_vcf.sh hipstr 1kg.snp.str.chr18.vcf.gz input_chr18_fortesting.vcf.gz converted_input_chr18_fortesting.vcf.gz`

...where 1kg.snp.str.chr18.vcf.gz was the reference panel downloaded from http://gymreklab.com/2018/03/05/snpstr_imputation.html, as mentioned above. I got another error here:

>Creating temporary file ... ~/tmp/tmp.ULy8Ry1uQ1.vcf
Copying over meta header lines from the reference panel, then copying the imputed file contents
bgzipping and tabix indexing
Adding INFO fields and values from the ref_panel, then removing loci which are missing required INFO fields
[W::hts_idx_load3] The index file is older than the data file: ~/1kg.snp.str.chr18.vcf.gz.tbi
The INFO tag "START" is not defined in ~/1kg.snp.str.chr18.vcf.gz, was the -h option provided?
Failed to read from standard input: unknown file type

Is this error related to the fact that there is no INFO tag of "START" (not sure what it is?) in the chr18 STR imputation reference file*, and/or something else related to the "unknown file type" (maybe my file is not in "HipSTR" format?)? 

Can you please help with this error? Thank you very much in advance!

Best,
Fahri

*PS: This is how the imputation reference file header and a randomly chosen STR look like:
>##fileformat=VCFv4.2
#filedate=20180225
##INFO=<ID=AF,Number=A,Type=Float,Description="Estimated   ALT Allele Frequencies">
##INFO=<ID=AR2,Number=1,Type=Float,Description="Allelic   R-Squared: estimated squared correlation between most probable REF dose and   true REF dose">
##INFO=<ID=DR2,Number=1,Type=Float,Description="Dosage   R-Squared: estimated squared correlation between estimated REF dose [P(RA) +   2*P(RR)] and true REF dose">
##INFO=<ID=IMP,Number=0,Type=Flag,Description="Imputed   marker">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=DS,Number=A,Type=Float,Description="estimated   ALT dose [P(RA) + P(AA)]">
##FORMAT=<ID=GP,Number=G,Type=Float,Description="Estimated   Genotype Probability">
##contig=<ID=18>

| #CHROM | POS    | ID         | REF                 | ALT                                                                                                                    | QUAL | FILTER | INFO                                                           | FORMAT | HG00096            |
|--------|--------|------------|---------------------|------------------------------------------------------------------------------------------------------------------------|------|--------|----------------------------------------------------------------|--------|--------------------|
| 18     | 147475 | STR_614123 | ATACAAAAAAAAAAAAAAA | ATACAAAAAAAAAAAAA,AAACAAAAAAAAAAAAAA,ATACAAAAAAAAAAAAAA,AAACAAAAAAAAAAAAAAA,ATACAAAAAAAAAAAAAAAA,ATACAAAAAAAAAAAAAAAAA | .    | PASS   | AR2=0.89;DR2=0.91;AF=0.016,0.0019,0.33,0.0025,0.0057,0.004;IMP | GT:DS  | 3\|0:0,0,0.99,0,0,0 |

### PR DESCRIPTION
**Pull request by gymreklab**
This PR introduces the `annotaTR` tool, which can be used to:
* Add dosages to TR VCF files
* Add Beagle annotations to imputed files (previously performed with `trtools_prep_beagle_vcf.sh`)
* Output to either VCF or PGEN format

It required adding pgenlib as a dependency and updating the version of cyvcf2 since we need `cyvcf2.VCF.num_variants`.

This feature is related to and should resolve #209. Steps to solve issues there:

```
# Perform TR imputation using the Saini panel
java -Xmx10g -jar beagle.19Apr22.7c0.jar \
    gt=1kg_snps_NA12878.vcf.gz \
    ref=1kg.snp.str.chr18.vcf.gz impute=true ap=true \
    out=1kg_imputed_chr18
tabix -p vcf 1kg_imputed_chr18.vcf.gz

# Annotate dosages and output both vcf and pgen
annotaTR \
    --vcf 1kg_imputed_chr18.vcf.gz --vcftype hipstr \
    --ref-panel 1kg.snp.str.chr18.vcf.gz \
    --outtype vcf pgen \
    --out test \
    --dosages bestguess_norm 
bgzip test.vcf
tabix -p vcf test.vcf.gz

# Test the pgen output is valid
plink2 --pfile test --validate

# Test the VCF output can be used as input to TRTools
statSTR --vcf test.vcf.gz --vcftype hipstr --afreq --out test
```

## Checklist

* [x ] I've checked to ensure there aren't already other open [pull requests](../../../pulls) for the same update/change
* [x ] I've prefixed the title of my PR according to [the conventional commits specification](https://www.conventionalcommits.org/). If your PR fixes a bug, please prefix the PR with `fix: `. Otherwise, if it introduces a new feature, please prefix it with `feat: `. If it introduces a breaking change, please add an exclamation before the colon, like `feat!: `. If the scope of the PR changes because of a revision to it, please update the PR title, since the title will be used in our CHANGELOG.
* [ x] At the top of the PR, I've [listed any open issues that this PR will resolve](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). For example, "resolves #0" if this PR resolves issue #0
- [ x] I've explained my changes in a manner that will make it possible for both users and maintainers of TRTools to understand them
* [ x] I've added tests for any new functionality. Or, if this PR fixes a bug, I've added test(s) that replicate it
* [ x] All directories with large test files are listed in [the "exclude" section](https://python-poetry.org/docs/pyproject/#include-and-exclude) of our pyproject.toml so that they do not appear in our PyPI distribution. All new files are also smaller than 0.5 MB.
* [ x] I've updated the relevant REAMDEs with any new usage information and checked that the newly built documentation is formatted properly
* [ x] All functions, modules, classes etc. still conform to [numpy docstring standards](https://numpydoc.readthedocs.io/en/latest/format.html)
* [x ] (if applicable) I've updated the pyproject.toml file with any changes I've made to TRTools's dependencies, and I've run `poetry lock --no-update` to ensure the lock file stays up to date and that our dependencies are locked to their minimum versions
* [ x] In the body of this PR, I've included a short address to the reviewer highlighting one or two items that might deserve their focus

### CODE REVIEW

**Review by aryarm — CHANGES_REQUESTED** [review]
Super exciting work! Thanks for taking this on, @mgymrek 

I took a look and have mostly just minor comments scattered throughout. Honestly, you gave so much attention to small edge cases and validating inputs in this PR, that I ultimately didn't feel like I had much to say, especially for the `annotaTR.py` script. So please let me know if there was anything in particular that you wanted me to look at that I might have missed.

### Code suggestions
- If you haven't already, it might be a good idea to confirm with @LiterallyUniqueLogin and @TaraMirmira that the `bestguess_norm` method will work wrt the clipping? I don't completely understand the clipping, but I think they had specific reasons for the clipping that might have been only relevant when working with Beagle dosages? Or if you're already testing that with @nicholema, then that's also fine.
- It might be a good idea to confirm that the PVAR file can be read by `cyvcf2`, since we will probably want to read from it in downstream tools. Perhaps this can be in a test?
    (Apparently, a VCF without a FORMAT column or any FORMAT fields is still a valid VCF, so the PVAR spec is actually a subset of the VCF spec!)
    Some of the PVAR files seem to have `None` instead of `.` and the INFO fields of each line aren't prefixed by `DSLEN=`
- _Not really a suggestion but just a note for the future_: Downstream users will probably prefer to get information from the PVAR file rather than the VCF, if they can help it. This is because the PVAR file doesn't have sample-specific info, so it will be much faster to read from it. ([See benchmarks for reading TRs from PVAR/PGEN files in haptools.](https://drive.google.com/file/d/1nYENunmcvw6NvA-RHgT1fsv7uCZP2Frx/view?usp=sharing)) Of course, someone could also just create a PVAR file with `plink2` and the VCF output of `annotaTR`, but I'm guessing it will be expensive for annotaTR to write that VCF and then they will also have that extra intermediate step:
    ```
    plink2 --make-just-pvar 'cols=vcfheader,qual,filter,info' --vcf annotaTR.vcf --out output --vcf-half-call m
    ```
    So perhaps we might one day want to create an additional `pvar` output type for annotaTR, so that we can write the equivalent of the VCF output but in a different, separate PVAR file? Unfortunately, I don't think there's an easy way to use `cyvcf2` to write a VCF without FORMAT fields, so I think this idea is dead in the water, but I figured it might be good to mention it anyway. Also, I don't think it's worth holding up this PR over something like that.

### Testing suggestions
The tests are pretty comprehensive for all different kinds of inputs, but I noticed that most of the tests only check the exit codes? It might also be good to check the outputs of annotaTR, as well. I know you already have tests for `GetDosages()`, but somebody might also try to change `annotaTR.py` one day, so we will probably want tests to check that the output of the command-line tool also stays the same. We did something like this in [test_mergeSTR.py](https://github.com/gymrek-lab/TRTools/blob/be63e79bb6402276e82cd7aea85bb76942217f75/trtools/mergeSTR/tests/test_mergeSTR.py#L277-L299), so you could probably copy/adapt that code.

While I was reading through the code, I also wrote down any edge cases I could think of. Here are a few that you might not already have tests for?
1. SNPs and TRs mixed into the file but no `--ref-panel`
2. non indexed VCF file input

And here are the edge cases that it seems like you already have tests for:
1. haploid calls
2. missing and low ploidy calls (-1 and -2)
3. A TR in the input VCF that's not in the ref panel
4. VCF output, PGEN output, both output
5. All of the different kinds of dosages
6. missing AP fields but beagle input requested
7. missing AP fields and no beagle input requested
8. no missing AP fields and no beagle input requested
9. no missing AP fields and beagle input requested
10. duplicate loci to ignore with `--ignore-duplicates`
11. all different kinds of reference panels: hipstr, advntr, gangstr, expansionhunter

File: pyproject.toml [hunk 1]
```diff
@@ -33,7 +33,7 @@ exclude = [
 [tool.poetry.dependencies]
 python = ">=3.7.1,<4.0"
 importlib-metadata = {version = ">=3.0.0", python = "<3.8"}
-cyvcf2 = ">=0.30.4"
+cyvcf2 = ">=0.31.1"
```
[the remark this answers is not in the document]
Comment (gymreklab) [reply pyproject.toml:36]: ok, thanks for looking into that!

File: trtools/annotaTR/README.rst [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  	  --dosages <string>
+
+where the argument to the :code:`--dosages` option specifies the method to compute dosages. It must be one of: 
+
+* :code:`bestguess`: dosages are computed by summing the length of alleles (given in number of repeat units) for each call. e.g. for a genotype heterozygous for 3 and 4 copies of a repeat, the best guess dosage will be 7.
+* :code:`bestguess_norm`: same as above, but scaled to be between 0 and 2. This is required if generating pgen output, since pgen only supports dosage values in this range.
+* :code:`beagpleap`: dosages for imputed calls are computed in a way that takes into account imputation uncertainty. This option requires the Beagle VCF FORMAT fields :code:`AP1` and :code:`AP2` to be present, which give the probability of each alternate allele on each of the two haplotypes of an individual. AP-based dosages are generated by computing a weighted sum of allele lengths for each haplotype (sum of alleleprob*allelelength over all alleles) and added together to get the total dosage for the genotype. In the case of imputation with no uncertainty (all AP fields are either 0 or 1), dosages computed in this way should match best guess dosages.
```
[the remark this answers is not in the document]
Comment (gymreklab) [reply trtools/annotaTR/README.rst:63]: nice!

File: poetry.lock
[the remark this answers is not in the document]
Comment (gymreklab) [reply poetry.lock:1]: thanks for updating those docs! I replaced the `poetry.lock` file with the one you suggested

File: trtools/annotaTR/annotaTR.py [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+        vcf_writer = cyvcf2.Writer(args.out+".vcf", reader)
+
+    # variant_ct needed for pgen
+    # If using a ref panel, assume we have same number
+    # of TRs as the panel
+    # Otherwise, assume our file is all TRs and use record count
+    if refpanel_metadata is not None:
+        variant_ct = ref_variant_ct
+    else:
+        variant_ct = reader.num_records
```
[the remark this answers is not in the document]
Comment (gymreklab) [reply trtools/annotaTR/annotaTR.py:512]: Yes this requires bgzip/index. This requirement was mentioned in the help message of annotaTR but not in the README so I added it. `checkgz` defaults to True for `utils.LoadSingleReader` so this should be caught but I added `checkgz=True` explicitly to make sure the requirement to have bgzip/index is more clear.

File: trtools/annotaTR/annotaTR.py [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+                        dosage_type=dosage_type, refreader=refreader):
+        common.WARNING("Error: problem initializing vcf header.")
+        return 1
+    if OutputFileTypes.vcf in outtypes:
+        vcf_writer = cyvcf2.Writer(args.out+".vcf", reader)
+
+    # variant_ct needed for pgen
+    # If using a ref panel, assume we have same number
+    # of TRs as the panel
+    # Otherwise, assume our file is all TRs and use record count
```
[the remark this answers is not in the document]
Comment (gymreklab) [reply trtools/annotaTR/annotaTR.py:504]: Good point! A mix of TRs and SNPs in the input file with no refpanel specified will result in an error when we try to convert a SNP record to a TRRecord. This would have resulted in a cryptic error before but I now added a check for this and will output a more helpful message before failing now:

```
"Error converting {chrom}:{pos} to a TR record. "
                "If your file is a mix of SNPs/TRs (e.g. from Beagle) you "
                "must provide a reference panel.".format(chrom=record.CHROM, pos=record.POS)
```

File: trtools/utils/utils.py [hunk 5]
```diff
@@ -18,7 +18,9 @@
 
 def LoadSingleReader(
         vcf_loc: str,
-        checkgz: bool = True) -> Optional[cyvcf2.VCF]:
+        checkgz: bool = True,
+        lazy: bool = False,
+        samples: List[str] = None) -> Optional[cyvcf2.VCF]:
```
[the remark this answers is not in the document]
Comment (gymreklab) [reply trtools/utils/utils.py:23]: Great suggestion! I have had similar hours for something else spent debugging after assuming sample order stays the same... I changed based on your suggestions.

(Here I just added it to avoid having cyvcf2 avoid loading any samples from the ref panel to save time, since we just need the INFO fields. but definitely this option could lead to problems down the road and set seems more appropriate)

File: trtools/utils/tr_harmonizer.py [hunk 6]
```diff
@@ -1065,6 +1083,93 @@ def UniqueStringGenotypes(self) -> Set[int]:
         """
         return set(self.UniqueStringGenotypeMapping().values())
 
+    def GetDosages(self, 
+            dosagetype: TRDosageTypes = TRDosageTypes.bestguess) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (aryarm) [reply trtools/utils/tr_harmonizer.py]: I just realized that `numpy.typing` requires `numpy>=1.20` and we only have `1.17` in our lock file

Feel free to ignore this suggestion in that case!

[the remark this answers is not in the document]
Comment (gymreklab) [reply trtools/utils/tr_harmonizer.py]: ok, will revert back to what we had

File: trtools/utils/tr_harmonizer.py [hunk 7]
```diff
@@ -1065,6 +1083,93 @@ def UniqueStringGenotypes(self) -> Set[int]:
...
+                "Requested Beagle dosages for record at {}:{} but AP1/AP2 fields not found.".format(self.chrom, self.pos)
+                )
+        unnorm_dosages = np.array([0]*self.GetNumSamples())
+
+        if dosagetype in [TRDosageTypes.bestguess, TRDosageTypes.bestguess_norm]:
+            # Get length gts and replace -1 (missing) and -2 (low ploidy) with 0
+            # But if normalizing set those to np.nan since unclear
+            # how to normalize haploid dosages which end up negative
+            # under the current method...
+            lengts = self.GetLengthGenotypes()
```
[the remark this answers is not in the document]
Comment (gymreklab) [reply trtools/utils/tr_harmonizer.py:1126]: Good point. Modified to return None if no samples, and added a test for this

File: trtools/annotaTR/annotaTR.py [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+from pgenlib import PgenWriter
+import sys
+
+import trtools.utils.common as common
+import trtools.utils.tr_harmonizer as trh
+import trtools.utils.utils as utils
+from trtools import __version__
+
+# Variables for PGEN output
+DEFAULT_PGEN_BATCHSIZE = 1000
```
[the remark this answers is not in the document]
Comment (gymreklab) [reply trtools/annotaTR/annotaTR.py:19]: Good point! Edited to expose as a command line option. Kept default 1000 since our own use cases have been primarily huge files

File: trtools/annotaTR/annotaTR.py [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+    for outtype in args.outtype:
+        try:
+            ot = OutputFileTypes[outtype]
+            outtypes.add(ot)
+        except KeyError:
+            common.WARNING("Invalid output type")
+            return 1
+    if args.vcftype != 'auto':
+        try:
+            checktype = trh.VcfTypes[args.vcftype]
```
[the remark this answers is not in the document]
Comment (gymreklab) [reply trtools/annotaTR/annotaTR.py:417]: correct this was just meant as a check to make sure it is valid input. modified to avoid creating a new variable

**Comment by gymreklab** [comment]
Thanks @aryam for the thorough review!

I have addressed the code suggestions and resolved all the conversations. See below regarding the other comments:

## Code suggestions

* Clipping/thresholding is to avoid rounding errors to ensure all the probabilities add up to 1
* Thanks for the suggestion regarding PVAR files as VCFs! I made some minor changes to make them valid (added QUAL/FILTER columns and fileformat header). Also added DSLEN prefix and changed "None" ID's to "."

## Testing suggestions

* I added a command line case for the case of SNPs/TRs mixed in but no ref-panel. This case would have failed since it would attempt to load the SNP as a TR record but now it fails more gracefully: e.g. "Error converting chr21:14245507 to a TR record. If your file is a mix of SNPs/TRs (e.g. from Beagle) you must provide a reference panel."
* Added command line tests for VCF files not being zipped or indexed

## Tests for files remaining the same

I addded tests based on the mergeSTR examples to check if files are unchanged. This required some minor changes to `trtools/testsupport/utils.py`:

* I added an option `max_lines_to_compare` to `assert_same_vcf`. Since I had to just keep the top lines of my example files otherwise they were too big. By default we should still read all lines of files to compare but for my test files I chopped them at 200 lines. (see `trtools/testsupport/sample_vcfs/annotaTR_vcfs/create_test_files.sh`)
* I modified `_make_info_dict` to not crash if an INFO field does not have a value. This is possible for INFO fields with `Number=0` like the `IMP` flag output by Beagle.

File: trtools/utils/tr_harmonizer.py [hunk 10]
```diff
@@ -1065,6 +1083,90 @@ def UniqueStringGenotypes(self) -> Set[int]:
...
+
+            # Get haplotype dosages
+            max_alt_len = max(self.alt_allele_lengths)
+            h1_dos = np.clip(np.dot(ap1, self.alt_allele_lengths), 0, max_alt_len)
+            h2_dos = np.clip(np.dot(ap2, self.alt_allele_lengths), 0, max_alt_len)
+            ref1_dos = ref1*self.ref_allele_length
+            ref2_dos = ref2*self.ref_allele_length
+
+            # Add together for final dosage
+            unnorm_dosages = (h1_dos + h2_dos + ref1_dos + ref2_dos).astype(np.float32)
```
Comment (aryarm) [inline trtools/utils/tr_harmonizer.py:1156]:
```suggestion
            unnorm_dosages = (h1_dos + h2_dos + ref1_dos + ref2_dos).astype(np.float32)
        else:
            raise ValueError("Unsupported dosagetype")
```

**Review by aryarm — APPROVED** [review]
ok, sounds good! Thanks for implementing my suggestions

And chopping at 200 lines sounds very reasonable. We shouldn't need more than a few lines, anyway.
